cpp/util/etcd.cc

```diff
@@ -607,16 +612,22 @@ Status EtcdClient::WatchState::HandleSingleValueRequestDone(
   }
   updates->emplace_back(status.ValueOrDie());
 
-  CHECK_LT(highest_index_seen_, updates->back().node_.modified_index_);
-  highest_index_seen_ =
-      max(highest_index_seen_, updates->back().node_.modified_index_);
-
   return Status::OK;
 }
 
 
 void EtcdClient::WatchState::RequestDone(Status status,
-                                         const shared_ptr<JsonObject>& json) {
+                                         const shared_ptr<JsonObject>& json,
+                                         int64_t etcd_index) {
+  // TODO(alcutter): doing this here works around some etcd 401 errors, but in
+  // the case of sustained high qps we could miss updates entirely (e.g. if
+  // this new index is already past the 1000 entry horizon by the time we make
+  // the new watch request.)  One way to address this might be to have the
+  // watcher re-do an "initial" get on the target, and, in the case of
+  // directory watches, maintain a set of known keys so that it can synthesise
+  // 'delete' updates.
+  highest_index_seen_ = max(highest_index_seen_, etcd_index);
```

Comment:
Yah, as commented IRL, I'll make it `CHECK_LE` here because I'm not 100% sure that there couldn't be some sort of etcd spasm where we get multiple `RequestDone()` coming back at the same index. But as you say, if someone's going to violate causality, they can do it somewhere else.

```
We don't allow faster than light neutrinos in here, said the bartender.
A neutrino walks into a bar.
```